keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+After this stage format and content of json formatted logs will be still in pre-alpha state and most likely will change after manual review.
+After completing this stage adding **new log entries should require using only the new logr-based logging API**. Double writing using both klog and logr APIs for new log entries should not be needed.
+
+#### Preparation
+* Develop and merge to Kubernetes master 2 implementations/configurations of logr API which will be needed in the latter steps of the migration:
+    * `no-op` implementation which does not log anything,
+    * `json` log format implementation,
+* Implement `k8s.io/component-base/logging` package which during its initialization should read `--logging-format` flag and initialize two global variables providing access to:
+    * logr implementation to be used to migrate existing logs (`json` logger for `json` log format and `no-op` logger for legacy format)
+    * logr implementation to be used by new logs (`json` loggers for `json` log format and klogr logger for legacy format)
+* For each package in Kubernetes source code define a logging module to which it will belong. Each logging module can span across multiple packages.
+* For each logging module define two variables with logr implementations to be used for existing and new logs initialized with the name of the module. Those should only be used until stage 2 of the migration is fully completed.
+* Implement a package implementing klog API which would check a global variable and either passthrough call to klog (with a proper deep value handing) or do nothing. This global variable should be initialized based on `--logging-format` value.
+* Kubernetes and all of its dependencies should be upgraded to klog v2.
+
+#### Automatic migration
+
+1. If `json` log format is selected klog v2 library should be initialized with json logr implementation so that all logs produced by Kubernetes or its dependencies using klog API would be redirected to logr.
+1. For all golang packages in Kubernetes following changes should be performed:
+    1. All imports and usages of klog package should be replaced with newly created klog wrapper.
+    1. For all calls to klog.V().Info*(), klog.Info*(), klog.Warning*(), klog.Error*(), klog.Fatal*() or klog.Exit*() corresponding logr invocation should be added just after those.
+
+This should be done automatically for the whole source code of the Kubernetes by a tool which would analyze the source code and edit it accordingly. The result of the automated migration should be verified and where necessary cleaned up manually.
+Generated PRs should be divided by directories and send for review.
+
+In case where constant string is passed as a message argument to logr.Info() or logr.Error() it should be replaced with a normalized constant string produced by removing all formatting placeholders like %s, %q and %v and removing any trailing non alphanumeric characters from its end:
```

Comment:
do we need future PR checkers to ensure we do not regress in the future?

Reply:
Yes, having a PR checker was our primary idea to enforce this.